components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginStorageDelegate.kt

```diff
@@ -0,0 +1,146 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.storage.Login
+import mozilla.components.concept.storage.LoginStorageDelegate
+import mozilla.components.lib.dataprotect.SecureAbove22Preferences
+
+/**
+ * A type of persistence operation, either 'create' or 'update'.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+enum class Operation { CREATE, UPDATE }
+
+internal const val PASSWORDS_KEY = "passwords"
+
+/**
+ * [LoginStorage.Delegate] implementation.
+ *
+ * An abstraction that handles the persistence and retrieval of [LoginEntry]s so that Gecko doesn't
+ * have to.
+ *
+ * In order to use this class, attach it to the active [GeckoRuntime] as its `loginStorageDelegate`.
+ * It is not designed to work with other engines.
+ */
+class DefaultLoginStorageDelegate(
+    private val loginStorage: AsyncLoginsStorage,
+    keyStore: SecureAbove22Preferences,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+) : LoginStorageDelegate {
+
+    private val password = { scope.async { keyStore.getString(PASSWORDS_KEY)!! } }
+
+    override fun onLoginUsed(login: Login) {
+        val guid = login.guid
+        if (guid == null || guid.isEmpty()) return
+        scope.launch {
+            loginStorage.withUnlocked(password) {
+                loginStorage.touch(guid).await()
+            }
+        }
+    }
+
+    override fun onLoginFetch(domain: String): List<Login> {
+        return runBlocking {
+            // GV expects a synchronous response. Blocking here hasn't caused problems during
+            // testing, but we should keep an eye on telemetry
+            loginStorage.withUnlocked(password) {
+                loginStorage.getByBaseDomain(domain).await()
+                    .map { it.toLogin() }
+            }
+        }
+    }
+
+    @Synchronized
+    override fun onLoginSave(login: Login) {
+        scope.launch {
+            loginStorage.withUnlocked(password) {
+                val existingLogin = login.guid?.let { loginStorage.get(it) }?.await()
+
+                when (getPersistenceOperation(login, existingLogin)) {
+                    Operation.UPDATE -> {
+                        existingLogin?.let { loginStorage.update(it.mergeWithLogin(login)).await() }
+                    }
+                    Operation.CREATE -> {
+                        loginStorage.add(login.toServerPassword()).await()
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Returns whether an existing login record should be [UPDATE]d or a new one [CREATE]d, based
+ * on the saved [ServerPassword] and new [Login].
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+fun getPersistenceOperation(newLogin: Login, savedLogin: ServerPassword?): Operation = when {
+    newLogin.guid.isNullOrEmpty() || savedLogin == null -> Operation.CREATE
```

Comment:
You have a funny third option: `savedLogin != null && newLogin.guid != savedLogin.guid`, which would indicate a bug somewhere... Maybe worth at least logging that.